src/services/completions.ts

```diff
@@ -1529,6 +1541,25 @@ namespace ts.Completions {
             });
         }
 
+        function getVariableDeclaration(property: Node): VariableDeclaration | undefined {
+            if (!isIdentifier(property)) {
+                return undefined;
+            }
+
+            const variableDeclaration = findAncestor(property, (node) => {
+                if (isFunctionLikeDeclaration(node)) {

```

Comment:
You may want to check if you’re actually in the body of the function, since this fails to filter out e.g. parameter defaults:

```ts
const fn = (p = /* 'fn' appears here */) => {}
```

That said, leaving in some bad completions that we already have is not a big problem IMO, so this is a minor nitpick.

A similar comment applies to `isVariableDeclaration(node)`: what you’re interested in is whether we walked up from the _initializer_ of a variable declaration, but you might have a false positive on a case like this:

```ts
const { a, b = /**/ } = { ... }
```

We’re not interested in getting the variable declaration if we started at the default binding for `b`. But, it ends up having no effect on the outcome, since there will be no symbol where `symbol.valueDeclaration === variableDeclaration`. But if it’s easy to skip that property access and equality check, we should do it, since it runs for every completion item. It could maybe shave off a couple milliseconds if the list is long.

This also brings up another point worth noting, that this logic won’t work for destructuring:

```ts
const { a, b } = { /** }; // 'a' and 'b' are offered
```

but I don’t think it’s worth doing anything about this. I really don’t want to have to loop through every binding element declaration for every symbol—I’m happy that currently the check is only one equality comparison per symbol.

Reply:
1. `You may want to check if you’re actually in the body of the function, since this fails to filter out e.g. parameter defaults:`
Yes. You are right, maybe we should use `isFunctionBlock` to check.